File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        h = { nodeVal: i for i, nodeVal in enumerate(inorder)}


        def b(p, l, r):
            if l > r or not p:
                return None
                # or not p
            val = p.popleft()
            root = TreeNode(val)
            idx = h[val]
            le = b(p, l, idx-1)
            ri = b(p, idx+1, r)

            root.left = le
            root.right = ri
            return root

        return b(deque(preorder), 0, len(inorder)-1)

        

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.java
class Solution {
    // 1. Global counter to track position in preorder array
    int pi = 0; 
    
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        pi = 0; // Important: Reset for every new test case
        Map<Integer, Integer> ino = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) {
            ino.put(inorder[i], i);
        }
        
        return f(preorder, 0, inorder.length - 1, ino);
    }

    private TreeNode f(int[] preorder, int l, int r, Map<Integer, Integer> mp) {
        // Base case: no elements left in this subtree range
        if (l > r) return null;

        // 2. Use the counter to get the current root value
        int val = preorder[pi];
        pi++; // Move to the next value for the next recursive call

        TreeNode root = new TreeNode(val);

        // 3. Split using the inorder map
        int inId = mp.get(val);

        // 4. Recursive calls (Order matters: Left then Right)
        root.left = f(preorder, l, inId - 1, mp);
        root.right = f(preorder, inId + 1, r, mp);

        return root;
    }
}